lib/THC/THCGeneral.c

```diff
@@ -3,56 +3,121 @@
+
+  /* Allocator using cudaMallocHost. */
+  THAllocator* cudaHostAllocator;
+  THCDeviceAllocator cudaDeviceAllocator;
+
+  /* Index of the current selected per-device resource. Actual CUDA resource
+     changes based on the current device, since resources are per-device */
+  THCThreadLocal/*<int>*/ currentPerDeviceStream;
+  THCThreadLocal/*<int>*/ currentPerDeviceBlasHandle;
+
+  /* Table of enabled peer-to-peer access between directed pairs of GPUs.
+     If i accessing allocs on j is enabled, p2pAccess[i][j] is 1; 0 otherwise. */
+  int** p2pAccessEnabled;
+
+  /* Is direct cross-kernel p2p access allowed? Normally, only cross-GPU
+     copies are allowed via p2p if p2p access is enabled at all for
+     the pair of GPUs in question, but if this flag is true, then
+     all cross-GPU access checks are disabled, allowing kernels to
+     directly access memory on another GPUs.
+     Note that p2p access must exist and be enabled for the pair of
+     GPUs in question. */
+  int p2pKernelAccessEnabled;
+
+  void (*cutorchGCFunction)(void *data);
+  void *cutorchGCData;
+  long heapSoftmax;
+  long heapDelta;
+};
+
 THCCudaResourcesPerDevice* THCState_getDeviceResourcePtr(
   THCState *state, int device);
 
 static void THCState_initDefaultDeviceAllocator(THCDeviceAllocator* a);
 
+THCState* THCState_alloc()
+{
+  THCState* state = (THCState*) malloc(sizeof(THCState));
+  memset(state, 0, sizeof(THCState));
+  return state;
+}
+
+void THCState_free(THCState* state)
+{
+  free(state);
+}
+
 void THCudaInit(THCState* state)
 {
-  state->cutorchGCFunction = NULL;
-  state->cutorchGCData = NULL;
   if (!state->cudaDeviceAllocator.malloc) {
     THCState_initDefaultDeviceAllocator(&state->cudaDeviceAllocator);
   }
 
-  int count = 0;
-  THCudaCheck(cudaGetDeviceCount(&count));
+  int numDevices = 0;
+  THCudaCheck(cudaGetDeviceCount(&numDevices));
+  state->numDevices = numDevices;
 
   int device = 0;
   THCudaCheck(cudaGetDevice(&device));
 
-  state->rngState = (THCRNGState*)malloc(sizeof(THCRNGState));
-  THCRandom_init(state, count, device);
+  /* Start in the default stream on the current device */
+  state->currentPerDeviceStream = THCThreadLocal_alloc();
+  state->currentPerDeviceBlasHandle = THCThreadLocal_alloc();
 
-  THCAllocator_init(state);
+  state->resourcesPerDevice = (THCCudaResourcesPerDevice*)
+    malloc(numDevices * sizeof(THCCudaResourcesPerDevice));
+  memset(state->resourcesPerDevice, 0, numDevices * sizeof(THCCudaResourcesPerDevice));
 
-  state->numDevices = count;
   state->deviceProperties =
-    (struct cudaDeviceProp*)malloc(count * sizeof(struct cudaDeviceProp));
+    (struct cudaDeviceProp*)malloc(numDevices * sizeof(struct cudaDeviceProp));
 
-  state->numUserStreams = 0;
-  state->numUserBlasHandles = 0;
+  state->rngState = (THCRNGState*)malloc(sizeof(THCRNGState));
+  THCRandom_init(state, numDevices, device);
+
+  state->cudaHostAllocator = (THAllocator*)malloc(sizeof(THAllocator));
```

Comment:
(you could make it a static struct that lives in THCAllocator.c, but this is all unrelated changes to THC. The short answer is that this is how it was before this PR)